pennylane/compiler/python_compiler/transforms/quantum/tree_traversal.py

```diff
@@ -0,0 +1,1587 @@
+
+        if "qnode" not in func_op.attributes:
+            return
+
+        self.module = self.get_parent_of_type(func_op, builtin.ModuleOp)
+        assert self.module is not None, "got orphaned qnode function"
+
+        # If no measurements found, no need to apply tree-traversal
+        if not self.check_if_qnode_have_mcm(func_op):
+            return
+
+        # Start with creating a new QNode function that will perform the tree traversal simulation.
+        # We prep the original QNode by ensuring measure boundaries are also register boundaries.
+        self.simple_io_func = self.simplify_quantum_io(func_op, rewriter)
+
+        self.setup_traversal_function(self.simple_io_func, rewriter)
+
+        self.split_traversal_segments(self.simple_io_func, rewriter)
+
+        self.initialize_data_structures(rewriter)
+
+        self.generate_traversal_code(rewriter)
+
+        self.finalize_traversal_function(rewriter)
+
+    def get_idx(self, op: Operation) -> int | None:
+        """Get the index of the operation."""
+        return op.idx if op.idx else op.idx_attr
+
+    def check_if_qnode_have_mcm(self, func_op: func.FuncOp) -> bool:
+        """Check if the QNode function contains any measurement operations."""
+        op_walker = func_op.body.walk()
+        for op in op_walker:
+            if isinstance(op, quantum.MeasureOp):
+                return True
+        return False
+
+    def simplify_quantum_io(self, func_op: func.FuncOp, rewriter: PatternRewriter) -> func.FuncOp:
+        """In order to facilitate quantum value handling, we will reinsert all extracted qubits
+        into the register at the end of each segment, and only allow the register as quantum
+        input and output of segments.
+
+        This pass guarantees that each measure op is preceded by exactly 1 ExtractOp, and whose
+        input is a fully "reassembled" register ready to be passed across difficult program
+        boundaries (e.g. control flow, function calls).
+        """
+        cloned_fun = func_op.clone()
+        cloned_fun.sym_name = builtin.StringAttr(func_op.sym_name.data + ".simple_io")
+        rewriter.insert_op(cloned_fun, InsertPoint.after(func_op))
+
+        current_reg = None
+        qubit_to_reg_idx = {}
+
+        for op in cloned_fun.body.ops:
+            match op:
+                case quantum.AllocOp():
+                    current_reg = op.qreg
+                case quantum.ExtractOp():
+                    # update register since it might have changed
+                    extract_idx = self.get_idx(op)
+                    qubit_to_reg_idx[op.qubit] = extract_idx
+                    op.operands = (current_reg, extract_idx)
+                case quantum.CustomOp():
+                    for i, qb in enumerate(chain(op.in_qubits, op.in_ctrl_qubits)):
+                        qubit_to_reg_idx[op.out_qubits[i]] = i
+                        qubit_to_reg_idx[op.results[i]] = qubit_to_reg_idx[qb]
```

Comment:
What's the point of the first insertion into `qubit_to_reg_idx`? Doesn't the second one make the first one redundant?